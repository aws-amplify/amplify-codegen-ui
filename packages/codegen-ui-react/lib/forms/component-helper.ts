/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

import { FieldConfigMetadata, StudioDataSourceType, StudioFormActionType } from '@aws-amplify/codegen-ui';
import { BinaryExpression, factory, Identifier, JsxAttribute, SyntaxKind } from 'typescript';
import { resetValuesName } from './form-state';
import { FIELD_TYPE_TO_TYPESCRIPT_MAP } from './typescript-type-map';

export const ControlledComponents = ['StepperField', 'SliderField', 'SelectField', 'ToggleButton', 'SwitchField'];

/**
 * given the component returns true if the component is a controlled component
 *
 * @param componentType
 * @returns
 */
export const isControlledComponent = (componentType: string): boolean => ControlledComponents.includes(componentType);

export const convertedValueAttributeMap: Record<string, (valueIdentifier: Identifier) => BinaryExpression> = {
  AWSTimestamp: (valueIdentifier: Identifier) =>
    factory.createBinaryExpression(
      valueIdentifier,
      factory.createToken(SyntaxKind.AmpersandAmpersandToken),
      factory.createCallExpression(factory.createIdentifier('convertToLocal'), undefined, [
        factory.createCallExpression(factory.createIdentifier('convertTimeStampToDate'), undefined, [valueIdentifier]),
      ]),
    ),
  AWSDateTime: (valueIdentifier: Identifier) =>
    factory.createBinaryExpression(
      valueIdentifier,
      factory.createToken(SyntaxKind.AmpersandAmpersandToken),
      factory.createCallExpression(factory.createIdentifier('convertToLocal'), undefined, [
        factory.createNewExpression(factory.createIdentifier('Date'), undefined, [valueIdentifier]),
      ]),
    ),
};

/**
 *
 * @param stateName name of the state name to render for the default value
 * @param { dataType } the dataType
 * @returns
 */
export const renderDefaultValueAttribute = (stateName: string, { dataType }: FieldConfigMetadata) => {
  const identifier = factory.createIdentifier(stateName);
  let expression = factory.createJsxExpression(undefined, identifier);

  if (dataType && typeof dataType !== 'object' && convertedValueAttributeMap[dataType]) {
    expression = factory.createJsxExpression(undefined, convertedValueAttributeMap[dataType](identifier));
  }

  return factory.createJsxAttribute(factory.createIdentifier('defaultValue'), expression);
};

export const renderValueAttribute = ({
  fieldConfig,
  componentName,
  currentValueIdentifier,
}: {
  fieldConfig: FieldConfigMetadata;
  componentName: string;
  currentValueIdentifier?: Identifier;
}): JsxAttribute | undefined => {
  const componentType = fieldConfig.studioFormComponentType ?? fieldConfig.componentType;
  const shouldGetForUncontrolled = fieldConfig.isArray;

  const valueIdentifier = currentValueIdentifier || getValueIdentifier(componentName, componentType);

  const controlledComponentToAttributesMap: { [key: string]: JsxAttribute } = {
    ToggleButton: factory.createJsxAttribute(
      factory.createIdentifier('isPressed'),
      factory.createJsxExpression(undefined, valueIdentifier),
    ),
    SliderField: factory.createJsxAttribute(
      factory.createIdentifier('value'),
      factory.createJsxExpression(undefined, valueIdentifier),
    ),
    SelectField: factory.createJsxAttribute(
      factory.createIdentifier('value'),
      factory.createJsxExpression(undefined, valueIdentifier),
    ),
    StepperField: factory.createJsxAttribute(
      factory.createIdentifier('value'),
      factory.createJsxExpression(undefined, valueIdentifier),
    ),
    SwitchField: factory.createJsxAttribute(
      factory.createIdentifier('isChecked'),
      factory.createJsxExpression(undefined, valueIdentifier),
    ),
    CheckboxField: factory.createJsxAttribute(
      factory.createIdentifier('checked'),
      factory.createJsxExpression(undefined, valueIdentifier),
    ),
  };

  if (controlledComponentToAttributesMap[componentType]) {
    return controlledComponentToAttributesMap[componentType];
  }

  // TODO: all components should be controlled once conversions are solid
  if (shouldGetForUncontrolled) {
    const { dataType } = fieldConfig;

    let expression = factory.createJsxExpression(undefined, valueIdentifier);

    if (dataType && typeof dataType !== 'object' && convertedValueAttributeMap[dataType]) {
      expression = factory.createJsxExpression(undefined, convertedValueAttributeMap[dataType](valueIdentifier));
    }

    return factory.createJsxAttribute(factory.createIdentifier('value'), expression);
  }

  return undefined;
};

/**
 * returns resetFunctionCheck Expression if a create form is datastore enabled
 *
 * @param studioForm {StudioForm}
 * @returns
 */
export const resetFunctionCheck = ({
  dataSourceType,
  formActionType,
}: {
  dataSourceType: StudioDataSourceType;
  formActionType: StudioFormActionType;
}) => {
  if (dataSourceType === 'DataStore' && formActionType === 'create') {
    return [
      factory.createIfStatement(
        factory.createIdentifier('clearOnSuccess'),
        factory.createBlock(
          [factory.createExpressionStatement(factory.createCallExpression(resetValuesName, undefined, []))],
          true,
        ),
        undefined,
      ),
    ];
  }
  return [];
};

const getValueIdentifier = (componentName: string, componentType: string) => {
  // For Boolean components like SwitchField, they need the full dot notation name
  // e.g. isChecked={options.enabled}
  if (FIELD_TYPE_TO_TYPESCRIPT_MAP[componentType] === SyntaxKind.BooleanKeyword) {
    return factory.createIdentifier(componentName);
  }
  return factory.createIdentifier(componentName.split('.')[0]);
};
