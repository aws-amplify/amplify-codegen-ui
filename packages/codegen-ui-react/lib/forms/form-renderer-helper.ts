/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

import { StateStudioComponentProperty, StudioForm, StudioFormActionType } from '@aws-amplify/codegen-ui';
import { BindingElement, factory, NodeFlags, PropertySignature, SyntaxKind } from 'typescript';
import { ImportCollection, ImportSource, ImportValue } from '../imports';
import { getStateName, getSetStateName } from '../react-component-render-helper';
import { addSchemaToArguments, getActionHookImportValue, getActionIdentifier } from '../workflow';

export const FormTypeDataStoreMap: Record<StudioFormActionType, string> = {
  create: 'Amplify.DataStoreCreateItemAction',
  update: 'Amplify.DataStoreUpdateItemAction',
};

export const FieldStateVariable = (componentName: string): StateStudioComponentProperty => ({
  componentName,
  property: 'fields',
});

/**
 * - formFields
 */
export const buildFieldStateStatements = (formName: string, importCollection: ImportCollection) => {
  importCollection.addMappedImport(ImportValue.USE_STATE_MUTATION_ACTION);

  return factory.createVariableStatement(
    undefined,
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createArrayBindingPattern([
            factory.createBindingElement(
              undefined,
              undefined,
              factory.createIdentifier(getStateName(FieldStateVariable(formName))),
              undefined,
            ),
            factory.createBindingElement(
              undefined,
              undefined,
              factory.createIdentifier(getSetStateName(FieldStateVariable(formName))),
              undefined,
            ),
          ]),
          undefined,
          undefined,
          factory.createCallExpression(factory.createIdentifier('useStateMutationAction'), undefined, [
            factory.createObjectLiteralExpression(),
          ]),
        ),
      ],
      NodeFlags.Const,
    ),
  );
};
/**
 *
 * @param form StudioForm
 * @param importCollection ImportCollection
 * @returns ActionStatement
 * renders the state variable for datastore and adds imports
 *
 * ex. for form create
 * const myFormonSubmit = useDataStoreCreateAction({
 *  model: myModel,
 *  fields: myFormFields,
 *  schema: schema
 * });
 */
export const buildDataStoreActionStatement = (form: StudioForm, importCollection: ImportCollection) => {
  const {
    dataType: { dataTypeName },
    formActionType,
  } = form;
  const actionHookImportValue = getActionHookImportValue(FormTypeDataStoreMap[formActionType]);
  importCollection.addMappedImport(actionHookImportValue);
  importCollection.addImport(ImportSource.LOCAL_MODELS, dataTypeName);
  const properties = [
    // model name
    factory.createPropertyAssignment(factory.createIdentifier('model'), factory.createIdentifier(dataTypeName)),
    // fields object name
    factory.createPropertyAssignment(
      factory.createIdentifier('fields'),
      factory.createIdentifier(getStateName(FieldStateVariable(form.name))),
    ),
  ];
  if (formActionType === 'update') {
    properties.push(factory.createPropertyAssignment(factory.createIdentifier('id'), factory.createIdentifier('id')));
  }
  addSchemaToArguments(properties, importCollection);

  return factory.createVariableStatement(
    undefined,
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createIdentifier(getActionIdentifier(form.name, 'onSubmit')),
          undefined,
          undefined,
          factory.createCallExpression(factory.createIdentifier(actionHookImportValue), undefined, [
            factory.createObjectLiteralExpression(properties, false),
          ]),
        ),
      ],
      NodeFlags.Const,
    ),
  );
};

export const buildMutationBindings = (form: StudioForm) => {
  const {
    dataType: { dataSourceType },
    formActionType,
  } = form;
  const elements: BindingElement[] = [];
  if (dataSourceType === 'DataStore') {
    if (formActionType === 'update') {
      elements.push(factory.createBindingElement(undefined, undefined, factory.createIdentifier('id'), undefined));
    }
    elements.push(
      factory.createBindingElement(undefined, undefined, factory.createIdentifier('onSubmitBefore'), undefined),
      factory.createBindingElement(undefined, undefined, factory.createIdentifier('onSubmitComplete'), undefined),
    );
  } else {
    elements.push(
      factory.createBindingElement(
        undefined,
        factory.createIdentifier('onSubmit'),
        getActionIdentifier(form.name, 'onSubmit'),
        undefined,
      ),
    );
  }
  elements.push(factory.createBindingElement(undefined, undefined, factory.createIdentifier('onCancel'), undefined));
  return elements;
};

/*
  generate params in typed props
  - datastore (onSubmitBefore(fields) & onSubmitComplete({saveSuccessful, errorMessage}))
   - if update include id
  - custom (onSubmit(fields))
 */
export const buildFormPropNode = (form: StudioForm) => {
  const {
    dataType: { dataSourceType },
    formActionType,
  } = form;
  const propSignatures: PropertySignature[] = [];
  if (dataSourceType === 'DataStore') {
    if (formActionType === 'update') {
      propSignatures.push(
        factory.createPropertySignature(
          undefined,
          factory.createIdentifier('id'),
          undefined,
          factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
        ),
      );
    }
    propSignatures.push(
      factory.createPropertySignature(
        undefined,
        'onSubmitBefore',
        factory.createToken(SyntaxKind.QuestionToken),
        factory.createFunctionTypeNode(
          undefined,
          [
            factory.createParameterDeclaration(
              undefined,
              undefined,
              undefined,
              'fields',
              undefined,
              factory.createTypeReferenceNode(factory.createIdentifier('Record'), [
                factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
                factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
              ]),
              undefined,
            ),
          ],
          factory.createTypeReferenceNode(factory.createIdentifier('Record'), [
            factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
            factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
          ]),
        ),
      ),
      factory.createPropertySignature(
        undefined,
        'onSubmitComplete',
        factory.createToken(SyntaxKind.QuestionToken),
        factory.createFunctionTypeNode(
          undefined,
          [
            factory.createParameterDeclaration(
              undefined,
              undefined,
              undefined,
              factory.createObjectBindingPattern([
                factory.createBindingElement(
                  undefined,
                  undefined,
                  factory.createIdentifier('saveSuccessful'),
                  undefined,
                ),
                factory.createBindingElement(undefined, undefined, factory.createIdentifier('errorMessage'), undefined),
              ]),
              undefined,
              factory.createTypeLiteralNode([
                factory.createPropertySignature(
                  undefined,
                  'saveSuccessful',
                  undefined,
                  factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
                ),
                factory.createPropertySignature(
                  undefined,
                  'errorMessage',
                  factory.createToken(SyntaxKind.QuestionToken),
                  factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
                ),
              ]),
              undefined,
            ),
          ],
          factory.createKeywordTypeNode(SyntaxKind.VoidKeyword),
        ),
      ),
    );
  }
  if (dataSourceType === 'Custom') {
    propSignatures.push(
      factory.createPropertySignature(
        undefined,
        'onSubmit',
        undefined,
        factory.createFunctionTypeNode(
          undefined,
          [
            factory.createParameterDeclaration(
              undefined,
              undefined,
              undefined,
              'fields',
              undefined,
              factory.createTypeReferenceNode(factory.createIdentifier('Record'), [
                factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
                factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
              ]),
              undefined,
            ),
          ],
          factory.createKeywordTypeNode(SyntaxKind.VoidKeyword),
        ),
      ),
    );
  }
  // onCancel?: () => void
  propSignatures.push(
    factory.createPropertySignature(
      undefined,
      'onCancel',
      factory.createToken(SyntaxKind.QuestionToken),
      factory.createFunctionTypeNode(undefined, [], factory.createKeywordTypeNode(SyntaxKind.VoidKeyword)),
    ),
  );
  return factory.createTypeLiteralNode(propSignatures);
};

/**
 *  TODO
 * - form valid boolean
 * - error objects { hasErrror: boolean, errorMessage: string }
 */
export const buildValidationStateStatements = () => {};
