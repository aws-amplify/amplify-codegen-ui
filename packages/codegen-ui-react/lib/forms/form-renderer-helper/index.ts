/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
export { onSubmitValidationRun, buildUpdateDatastoreQuery, buildExpression } from './cta-props';

export { buildModelFieldObject } from './model-fields';

export { buildMutationBindings, buildOverrideOnChangeStatement } from './event-handler-props';

export { buildOverrideTypesBindings } from './type-helper';

export { buildResetValuesOnRecordUpdate, buildSetStateFunction, getLinkedDataName, getPropName } from './form-state';

export { buildValidations, runValidationTasksFunction } from './validation';

export { addFormAttributes } from './all-props';

export { mapFromFieldConfigs, isManyToManyRelationship } from './map-from-fieldConfigs';

export { buildRelationshipQuery, buildGetRelationshipModels } from './relationship';

export { shouldWrapInArrayField } from './render-checkers';

export { renderArrayFieldComponent } from './render-array-field';

export { getDecoratedLabel } from './label-decorator';
