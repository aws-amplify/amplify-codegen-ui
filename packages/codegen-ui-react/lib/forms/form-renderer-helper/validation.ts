/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import {
  Expression,
  factory,
  NodeFlags,
  SyntaxKind,
  ObjectLiteralElementLike,
  ObjectLiteralExpression,
  PropertyAssignment,
} from 'typescript';
import { FieldValidationConfiguration, ValidationTypes, isValidVariableName } from '@aws-amplify/codegen-ui';

export const createValidationExpression = (validationRules: FieldValidationConfiguration[] = []): Expression => {
  const validateExpressions = validationRules.map<ObjectLiteralExpression>((rule) => {
    const elements: ObjectLiteralElementLike[] = [
      factory.createPropertyAssignment(factory.createIdentifier('type'), factory.createStringLiteral(rule.type)),
    ];
    if (
      rule.type === ValidationTypes.CONTAINS ||
      rule.type === ValidationTypes.NOT_CONTAINS ||
      rule.type === ValidationTypes.END_WITH ||
      rule.type === ValidationTypes.START_WITH ||
      rule.type === ValidationTypes.BE_BEFORE ||
      rule.type === ValidationTypes.BE_AFTER
    ) {
      elements.push(
        factory.createPropertyAssignment(
          factory.createIdentifier('strValues'),
          factory.createArrayLiteralExpression(
            rule.strValues.map((value) => factory.createStringLiteral(value)),
            false,
          ),
        ),
      );
    } else if (
      rule.type === ValidationTypes.LESS_THAN_CHAR_LENGTH ||
      rule.type === ValidationTypes.GREATER_THAN_CHAR_LENGTH ||
      rule.type === ValidationTypes.LESS_THAN_NUM ||
      rule.type === ValidationTypes.GREATER_THAN_NUM ||
      rule.type === ValidationTypes.EQUAL_TO_NUM
    ) {
      elements.push(
        factory.createPropertyAssignment(
          factory.createIdentifier('numValues'),
          factory.createArrayLiteralExpression(
            rule.numValues.map((value) => factory.createNumericLiteral(value)),
            false,
          ),
        ),
      );
    }
    if (rule.validationMessage) {
      elements.push(
        factory.createPropertyAssignment(
          factory.createIdentifier('validationMessage'),
          factory.createStringLiteral(rule.validationMessage),
        ),
      );
    }
    return factory.createObjectLiteralExpression(elements, false);
  });

  return factory.createArrayLiteralExpression(validateExpressions, true);
};

/**
 * builds validation variable
 * for nested values it will mention the full path as that corresponds to the fields
 * this will also link to error messages
 *
 * const validations = { post_url: [{ type: "URL" }], 'user.status': [] };
 *
 * @param fieldConfigs
 * @returns
 */
export function buildValidations(validationsAssignments: PropertyAssignment[]) {
  return factory.createVariableStatement(
    undefined,
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createIdentifier('validations'),
          undefined,
          undefined,
          factory.createObjectLiteralExpression(validationsAssignments, true),
        ),
      ],
      NodeFlags.Const,
    ),
  );
}

export function buildValidationForField(fieldName: string, validationRules: FieldValidationConfiguration[]) {
  const propKey =
    fieldName.split('.').length > 1 || !isValidVariableName(fieldName)
      ? factory.createStringLiteral(fieldName)
      : factory.createIdentifier(fieldName);
  return factory.createPropertyAssignment(propKey, createValidationExpression(validationRules));
}

/**
    if (errors.name?.hasError) {
      runValidationTasks("name", value);
    }
   */
export function getOnChangeValidationBlock(fieldName: string) {
  return factory.createIfStatement(
    factory.createPropertyAccessChain(
      isValidVariableName(fieldName)
        ? factory.createPropertyAccessExpression(
            factory.createIdentifier('errors'),
            factory.createIdentifier(fieldName),
          )
        : factory.createElementAccessExpression(
            factory.createIdentifier('errors'),
            factory.createStringLiteral(fieldName),
          ),
      factory.createToken(SyntaxKind.QuestionDotToken),
      factory.createIdentifier('hasError'),
    ),
    factory.createBlock(
      [
        factory.createExpressionStatement(
          factory.createCallExpression(factory.createIdentifier('runValidationTasks'), undefined, [
            factory.createStringLiteral(fieldName),
            factory.createIdentifier('value'),
          ]),
        ),
      ],
      true,
    ),
    undefined,
  );
}

/**
  const runValidationTasks = async (fieldName, currentValue, getDisplayValue) => {
    const value = currentValue && getDisplayValue ? getDisplayValue(currentValue) : currentValue;
    let validationResponse = validateField(value, validations[fieldName]);
    const customValidator = fetchByPath(onValidate, fieldName);
    if (customValidator) {
      validationResponse = await customValidator(value, validationResponse);
    }
    setErrors((errors) => ({ ...errors, [fieldName]: validationResponse }));
    return validationResponse;
  };
   */

export const runValidationTasksFunction = factory.createVariableStatement(
  undefined,
  factory.createVariableDeclarationList(
    [
      factory.createVariableDeclaration(
        factory.createIdentifier('runValidationTasks'),
        undefined,
        undefined,
        factory.createArrowFunction(
          [factory.createModifier(SyntaxKind.AsyncKeyword)],
          undefined,
          [
            factory.createParameterDeclaration(
              undefined,
              undefined,
              undefined,
              factory.createIdentifier('fieldName'),
              undefined,
              undefined,
              undefined,
            ),
            factory.createParameterDeclaration(
              undefined,
              undefined,
              undefined,
              factory.createIdentifier('currentValue'),
              undefined,
              undefined,
              undefined,
            ),
            factory.createParameterDeclaration(
              undefined,
              undefined,
              undefined,
              factory.createIdentifier('getDisplayValue'),
              undefined,
              undefined,
              undefined,
            ),
          ],
          undefined,
          factory.createToken(SyntaxKind.EqualsGreaterThanToken),
          factory.createBlock(
            [
              factory.createVariableStatement(
                undefined,
                factory.createVariableDeclarationList(
                  [
                    factory.createVariableDeclaration(
                      factory.createIdentifier('value'),
                      undefined,
                      undefined,
                      factory.createConditionalExpression(
                        factory.createBinaryExpression(
                          factory.createIdentifier('currentValue'),
                          factory.createToken(SyntaxKind.AmpersandAmpersandToken),
                          factory.createIdentifier('getDisplayValue'),
                        ),
                        factory.createToken(SyntaxKind.QuestionToken),
                        factory.createCallExpression(factory.createIdentifier('getDisplayValue'), undefined, [
                          factory.createIdentifier('currentValue'),
                        ]),
                        factory.createToken(SyntaxKind.ColonToken),
                        factory.createIdentifier('currentValue'),
                      ),
                    ),
                  ],
                  NodeFlags.Const,
                ),
              ),
              factory.createVariableStatement(
                undefined,
                factory.createVariableDeclarationList(
                  [
                    factory.createVariableDeclaration(
                      factory.createIdentifier('validationResponse'),
                      undefined,
                      undefined,
                      factory.createCallExpression(factory.createIdentifier('validateField'), undefined, [
                        factory.createIdentifier('value'),
                        factory.createElementAccessExpression(
                          factory.createIdentifier('validations'),
                          factory.createIdentifier('fieldName'),
                        ),
                      ]),
                    ),
                  ],
                  NodeFlags.Let,
                ),
              ),
              factory.createVariableStatement(
                undefined,
                factory.createVariableDeclarationList(
                  [
                    factory.createVariableDeclaration(
                      factory.createIdentifier('customValidator'),
                      undefined,
                      undefined,
                      factory.createCallExpression(factory.createIdentifier('fetchByPath'), undefined, [
                        factory.createIdentifier('onValidate'),
                        factory.createIdentifier('fieldName'),
                      ]),
                    ),
                  ],
                  NodeFlags.Const,
                ),
              ),
              factory.createIfStatement(
                factory.createIdentifier('customValidator'),
                factory.createBlock(
                  [
                    factory.createExpressionStatement(
                      factory.createBinaryExpression(
                        factory.createIdentifier('validationResponse'),
                        factory.createToken(SyntaxKind.EqualsToken),
                        factory.createAwaitExpression(
                          factory.createCallExpression(factory.createIdentifier('customValidator'), undefined, [
                            factory.createIdentifier('value'),
                            factory.createIdentifier('validationResponse'),
                          ]),
                        ),
                      ),
                    ),
                  ],
                  true,
                ),
                undefined,
              ),
              factory.createExpressionStatement(
                factory.createCallExpression(factory.createIdentifier('setErrors'), undefined, [
                  factory.createArrowFunction(
                    undefined,
                    undefined,
                    [
                      factory.createParameterDeclaration(
                        undefined,
                        undefined,
                        undefined,
                        factory.createIdentifier('errors'),
                        undefined,
                        undefined,
                        undefined,
                      ),
                    ],
                    undefined,
                    factory.createToken(SyntaxKind.EqualsGreaterThanToken),
                    factory.createParenthesizedExpression(
                      factory.createObjectLiteralExpression(
                        [
                          factory.createSpreadAssignment(factory.createIdentifier('errors')),
                          factory.createPropertyAssignment(
                            factory.createComputedPropertyName(factory.createIdentifier('fieldName')),
                            factory.createIdentifier('validationResponse'),
                          ),
                        ],
                        false,
                      ),
                    ),
                  ),
                ]),
              ),
              factory.createReturnStatement(factory.createIdentifier('validationResponse')),
            ],
            true,
          ),
        ),
      ),
    ],
    NodeFlags.Const,
  ),
);
