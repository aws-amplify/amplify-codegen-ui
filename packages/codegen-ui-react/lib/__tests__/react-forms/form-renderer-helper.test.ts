/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { StudioForm, ValidationTypeMapping } from '@aws-amplify/codegen-ui';
import { EmitHint, Node } from 'typescript';
import { buildFormPropNode } from '../../forms';
import { buildPrinter, defaultRenderConfig } from '../../react-studio-template-renderer-helper';
import { createValidationExpression } from '../../forms/form-renderer-helper/validation';

describe('form-render utils', () => {
  let printNode: (node: Node) => string;

  beforeAll(() => {
    const { printer, file } = buildPrinter('myFileMock', defaultRenderConfig);
    printNode = (node: Node) => {
      return printer.printNode(EmitHint.Unspecified, node, file);
    };
  });

  it('should generate before & complete types if datastore config is set', () => {
    const form: StudioForm = {
      id: '123',
      name: 'mySampleForm',
      formActionType: 'create',
      dataType: { dataSourceType: 'DataStore', dataTypeName: 'Post' },
      fields: {},
      sectionalElements: {},
      style: {},
      cta: {},
    };

    const propSignatures = buildFormPropNode(form, {});
    const node = printNode(propSignatures);
    expect(node).toMatchSnapshot();
  });

  it('should generate regular onsubmit if dataSourceType is custom', () => {
    const form: StudioForm = {
      id: '123',
      name: 'myCustomForm',
      formActionType: 'create',
      dataType: { dataSourceType: 'Custom', dataTypeName: 'Custom' },
      fields: {},
      sectionalElements: {},
      style: {},
      cta: {},
    };
    const propSignatures = buildFormPropNode(form, {});
    const node = printNode(propSignatures);
    expect(node).toMatchSnapshot();
  });

  it('should render cancel props if included cancel object is an empty object', () => {
    const form: StudioForm = {
      id: '123',
      name: 'myCustomForm',
      formActionType: 'create',
      dataType: { dataSourceType: 'Custom', dataTypeName: 'Custom' },
      fields: {},
      sectionalElements: {},
      style: {},
      cta: {
        cancel: {},
      },
    };
    const propSignatures = buildFormPropNode(form, {});
    const node = printNode(propSignatures);
    expect(node).toContain('onCancel?: () => void;');
    expect(node).toMatchSnapshot();
  });

  it('should render composite primary keys', () => {
    const form: StudioForm = {
      id: '123',
      name: 'mySampleForm',
      formActionType: 'update',
      dataType: { dataSourceType: 'DataStore', dataTypeName: 'Post' },
      fields: {},
      sectionalElements: {},
      style: {},
      cta: {
        cancel: {},
      },
    };
    const propSignatures = buildFormPropNode(
      form,
      { description: { dataType: 'Int', validationRules: [], componentType: 'TextField' } },
      undefined,
      ['myKey', 'description'],
    );
    const node = printNode(propSignatures);
    expect(node).toContain('id?: {');
    expect(node).toContain('myKey: string;');
    expect(node).toContain('description: number;');
    expect(node).toMatchSnapshot();
  });

  test.each(
    ValidationTypeMapping.StringType.map<[string, any[]]>((type) => [
      `${type} with undefined numValues`,
      [
        {
          strValues: ['chard'],
          numValues: undefined,
          type,
        } as any,
      ],
    ]).concat(
      ValidationTypeMapping.NumberType.map<[string, any[]]>((type) => [
        `${type} with undefined strValues`,
        [
          {
            strValues: undefined,
            numValues: [9],
            type,
          } as any,
        ],
      ]),
    ),
  )('createValidationExpression handles %s', (description, rules) => {
    const expression = createValidationExpression(rules);
    const node = printNode(expression);
    expect(node).toMatchSnapshot();
  });
});
