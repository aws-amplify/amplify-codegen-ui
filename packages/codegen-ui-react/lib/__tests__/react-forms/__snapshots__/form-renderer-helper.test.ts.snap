// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form-render utils createValidationExpression handles BeAfter with undefined numValues 1`] = `
"[
    { type: \\"BeAfter\\", strValues: [\\"chard\\"] }
]"
`;

exports[`form-render utils createValidationExpression handles BeBefore with undefined numValues 1`] = `
"[
    { type: \\"BeBefore\\", strValues: [\\"chard\\"] }
]"
`;

exports[`form-render utils createValidationExpression handles Contains with undefined numValues 1`] = `
"[
    { type: \\"Contains\\", strValues: [\\"chard\\"] }
]"
`;

exports[`form-render utils createValidationExpression handles EndWith with undefined numValues 1`] = `
"[
    { type: \\"EndWith\\", strValues: [\\"chard\\"] }
]"
`;

exports[`form-render utils createValidationExpression handles EqualTo with undefined strValues 1`] = `
"[
    { type: \\"EqualTo\\", numValues: [9] }
]"
`;

exports[`form-render utils createValidationExpression handles GreaterThanChar with undefined strValues 1`] = `
"[
    { type: \\"GreaterThanChar\\", numValues: [9] }
]"
`;

exports[`form-render utils createValidationExpression handles GreaterThanNum with undefined strValues 1`] = `
"[
    { type: \\"GreaterThanNum\\", numValues: [9] }
]"
`;

exports[`form-render utils createValidationExpression handles LessThanChar with undefined strValues 1`] = `
"[
    { type: \\"LessThanChar\\", numValues: [9] }
]"
`;

exports[`form-render utils createValidationExpression handles LessThanNum with undefined strValues 1`] = `
"[
    { type: \\"LessThanNum\\", numValues: [9] }
]"
`;

exports[`form-render utils createValidationExpression handles NotContains with undefined numValues 1`] = `
"[
    { type: \\"NotContains\\", strValues: [\\"chard\\"] }
]"
`;

exports[`form-render utils createValidationExpression handles StartWith with undefined numValues 1`] = `
"[
    { type: \\"StartWith\\", strValues: [\\"chard\\"] }
]"
`;

exports[`form-render utils should generate before & complete types if datastore config is set 1`] = `
"{
    clearOnSuccess?: boolean;
    onSubmit?: (fields: mySampleFormInputValues) => mySampleFormInputValues;
    onSuccess?: (fields: mySampleFormInputValues) => void;
    onError?: (fields: mySampleFormInputValues, errorMessage: string) => void;
    onChange?: (fields: mySampleFormInputValues) => mySampleFormInputValues;
    onValidate?: mySampleFormValidationValues;
}"
`;

exports[`form-render utils should generate regular onsubmit if dataSourceType is custom 1`] = `
"{
    onSubmit: (fields: myCustomFormInputValues) => void;
    onChange?: (fields: myCustomFormInputValues) => myCustomFormInputValues;
    onValidate?: myCustomFormValidationValues;
}"
`;

exports[`form-render utils should render cancel props if included cancel object is an empty object 1`] = `
"{
    onSubmit: (fields: myCustomFormInputValues) => void;
    onCancel?: () => void;
    onChange?: (fields: myCustomFormInputValues) => myCustomFormInputValues;
    onValidate?: myCustomFormValidationValues;
}"
`;

exports[`form-render utils should render composite primary keys 1`] = `
"{
    id?: {
        myKey: string;
        description: number;
    };
    post?: Post;
    onSubmit?: (fields: mySampleFormInputValues) => mySampleFormInputValues;
    onSuccess?: (fields: mySampleFormInputValues) => void;
    onError?: (fields: mySampleFormInputValues, errorMessage: string) => void;
    onCancel?: () => void;
    onChange?: (fields: mySampleFormInputValues) => mySampleFormInputValues;
    onValidate?: mySampleFormValidationValues;
}"
`;
